src/hotspot/share/opto/cfgnode.hpp

```diff
@@ -452,6 +452,26 @@ class RangeCheckNode : public IfNode {
   virtual Node* Ideal(PhaseGVN *phase, bool can_reshape);
 };
 
+// Special node that denotes a Parse Predicate added during parsing. A Parse Predicate serves as placeholder to later
+// create Runtime Predicates above it. They all share the same uncommon trap. The Parse Predicate will follow the
+// Runtime Predicates. Together they form a Regular Predicate Block. There are three kinds of Parse Predicates:
+// Loop Parse Predicate, Profiled Loop Parse Predicate (both used by Loop Predication), and Loop Limit Check Parse
+// Predicate (used for integer overflow checks when creating a counted loop).
+// More information about predicates can be found at loopPredicate.cpp.
+class ParsePredicateNode : public IfNode {
+  Deoptimization::DeoptReason _deopt_reason;
```

Comment:
Don't we need to override `Node::hash()` and `Node::cmp()` here to account for the `_deopt_reason` field?

Reply:
Since it is a CFG node, the hash should always be different due to a different control input. If two `If` nodes have the same control input after an optimization, the graph is broken.

Reply:
Right, I missed that. Looks goot then!